### PR TITLE
chunk items to delete by 1000

### PR DESCRIPTION
**Pull request by alouvion-aneo**
AWS S3 API has a limit of 1000 items to delete in his batch delete mode.
So if we want to delete more than 1000 items we have to chunk by 1000 or less.

### CODE REVIEW

File: Adaptors/S3/src/ObjectStorage.cs [hunk 1]
```diff
@@ -169,17 +168,37 @@ await taskList.WhenAll()
...
-                                   };
+                                    });
 
-    var deleteObjectsResponse = await s3Client_.DeleteObjectsAsync(multiObjectDeleteRequest)
-                                               .ConfigureAwait(false);
-    return deleteObjectsResponse.DeletedObjects.Count == valuesCount + 1;
+    var deletedItem = 0;
+    try
+    {
+      var multiObjectDeleteRequest = new DeleteObjectsRequest
```
Comment (aneojgurhem) [inline Adaptors/S3/src/ObjectStorage.cs:176]: You need to create this object in the select.
You tried to use thread safe access for the add but not for the object list.

Comment (alouvion-aneo) [reply Adaptors/S3/src/ObjectStorage.cs:176]: indeed

Comment (alouvion-aneo) [reply Adaptors/S3/src/ObjectStorage.cs]: updated

File: Adaptors/S3/src/ObjectStorage.cs [hunk 2]
```diff
@@ -169,17 +168,36 @@ await taskList.WhenAll()
...
+    {
+      await keyList.Chunk(MaxAllowedKeysPerDelete)
+                   .Select(async chunkedKeyList =>
+                           {
+                             var multiObjectDeleteRequest = new DeleteObjectsRequest
+                                                            {
+                                                              BucketName = bucketName_,
+                                                              Objects    = null,
+                                                            };
+                             multiObjectDeleteRequest.Objects = chunkedKeyList.ToList();
```
Comment (aneojgurhem) [inline Adaptors/S3/src/ObjectStorage.cs:184]: maybe you should be able to do something similar to:
```suggestion
                                                              Objects    =  {
                                                                  chunkedKeyList
                                                              },
                                                            };
```

Comment (alouvion-aneo) [reply Adaptors/S3/src/ObjectStorage.cs]: it is already the case... now...

**Review by aneojgurhem — APPROVED** [review]

**Pull request merged**